Add doc for complex config values (#31010)
angular/angular: aio/content/guide/workspace-config.md

## Project tool configuration options

Architect is the tool that the CLI uses to perform complex tasks, such as compilation and test running.
Architect is a shell that runs a specified [builder](guide/glossary#builder) to perform a given task, according to a [target](guide/glossary#target) configuration.
You can define and configure new builders and targets to extend the CLI.
See [Angular CLI Builders](guide/cli-builder).

{@a default-build-targets}

### Default Architect builders and targets

Angular defines default builders for use with specific CLI commands, or with the general `ng run` command.

> The JSON schemas that the define the options and defaults for each of these default builders are collected in the [`@angular-devkit/build-angular`](https://github.com/angular/angular-cli/blob/7.0.x/packages/angular/cli/lib/config/schema.json) package.

Reviewer: ```suggestion
The JSON schemas that the define the options and defaults for each of these default builders are collected in the [`@angular-devkit/build-angular`](https://github.com/angular/angular-cli/blob/8.0.x/packages/angular/cli/lib/config/schema.json) package.
```

The schemas configure options for the following targets.

* app-shell
* browser
* dev-server
* extract-i18n
* karma
* protractor
* server

> * tslint

Reviewer 2: Isn't the target name `lint`?

Reviewer 3: The below is not clear.

What is in the list, targets or builders? for example `tslint` is a builder, `lint` is a target, `karma` is a builder while `test` is the target.

```
Angular defines default builders for use with specific CLI commands, or with the general ng run command. The JSON schemas that the define the options and defaults for each of these default builders are collected in the @angular-devkit/build-angular package. The schemas configure options for the following targets.

app-shell
browser
dev-server
extract-i18n
karma
protractor
server
lint
```

My suggestion would be to change it to

```
```
Angular defines default builders for use with specific CLI commands, or with the general ng run command. The JSON schemas that the define the options and defaults for each of these default builders are collected in the @angular-devkit/build-angular package. The schemas configure options for the following builders.

app-shell
browser
dev-server
extract-i18n
karma
protractor
server
tslint
```
```

The configurable options for a default or targeted build generally correspond to the options available for the [`ng build`](cli/build), [`ng serve`](cli/serve), and [`ng test`](cli/test) commands. For details of those options and their possible values, see the [CLI Reference](cli).

Some additional options can only be set through the configuration file, either by direct editing or with the [`ng config`](cli/config) command.

| OPTIONS PROPERTIES | DESCRIPTION |
| :------------------------- | :---------------------------- |

> | `fileReplacements`         | An object containing files and their compile-time replacements. |

Reviewer 3: `fileReplacements` and `stylePreprocessorOptions` doesn't seem to be explained here.

Also, I suggest to add a link example See more in `Styles and scripts configuration`

Author: Can you provide better explanations?

Reviewer 3: For the stylePreprocessorOptions, you can check https://github.com/angular/angular-cli/wiki/stories-global-styles

File replacements: is explained here https://angular.io/guide/build#configure-target-specific-file-replacements

Author: @alan-agius4  Can you provide more description and examples for these? Do they need their own sections, or are they also about styles/scripts?

Reviewer 3: They need their own sections, these are 2 options which are totally unrelated to styles/scripts options.

| `stylePreprocessorOptions` | An object containing option-value pairs to pass to style preprocessors. |
| `assets`                   | An object containing paths to static assets to add to the global context of the project. The default paths point to the project's icon file and its `assets` folder.  See more in [Project asset configuration](#asset-config). |
| `styles`                   | An object containing style files to add to the global context of the project. Angular CLI supports CSS imports and all major CSS preprocessors: [sass/scss](http://sass-lang.com/), [less](http://lesscss.org/), and [stylus](http://stylus-lang.com/). |
| `scripts`                  | An object containing JavaScript script files to add to the global context of the project. The scripts are loaded exactly as if you had added them in a `<script>` tag inside `index.html`. |
| `budgets`                  | Default size-budget type and threshholds for all or parts of your app. You can configure the builder to report a warning or an error when the output reaches or exceeds a threshold size. See [Configure size budgets](guide/build#configure-size-budgets). (Not available in `test` section.) |

> ## Complex build configuration values

Reviewer 3: Complex configuration values

Reviewer 3: As mentioned below these options are not just for `ng build`

> Some build options that can be set to a simple Boolean value with a command flag, such as `sourceMap` and `optimization` can be given a complex value using the configuration file.

Reviewer 3: `Some build options` -> `Some options`

  { "glob": "favicon.ico", "input": "src/", "output": "/" }
]

</code-example>

You can use this extended configuration to copy assets from outside your project.
For example, the following configuration copies assets from a node package:

<code-example format="." language="json" linenums="false">

"assets": [
 { "glob": "**/*", "input": "./node_modules/some-package/images", "output": "/some-package/" },
]

</code-example>

The contents of `node_modules/some-package/images/` will be available in `dist/some-package/`.

The following example uses the `ignore` field to exclude certain files in the assets folder from being copied into the build:

<code-example format="." language="json" linenums="false">

"assets": [
 { "glob": "**/*", "input": "src/assets/", "ignore": ["**/*.svg"], "output": "/assets/" },
]

</code-example>

{@a style-config}

### Styles and scripts configuration

An array entry for the `styles` and `scripts` options can be a simple file references, or an object that points to an extra entry-point file.
The associated builder will load that file and its dependencies as a separate bundle during the build.

> Using this configuration you have the option of naming the bundle for the entry point, which enables it to be loaded lazily.

Reviewer 3: There is quite a misconception on what this `lazy` option mean.

In reality this doesn’t allow a script to be lazy loaded, but rather it will not be injected, which if than it needs to be lazy loaded, it will he up to the user to implement the logic to lazy load the script.

Last week we actually had a small chat around changing the name of this option to avoid confusion.

@clydin, shall we go forward with the suggested that you made last week and update this doc based on that?

Reviewer 4: Personally, I think we should as it appears to be an ongoing point of confusion.  There are also potential use cases beyond lazy loading for the option.
However, I think it should be discussed in the next status meeting to ensure we arrive at a consensus.

Reviewer 3: Created an issue https://github.com/angular/angular-cli/issues/14814

Author: Does this work to describe the (badly-named) option for now?

The bundle is injected by default, but you can set `lazy` to true to exclude the bundle from injection.
For example, the following object values create and name a bundle that contains styles and scripts, and excludes it from injection:

<code-example format="." language="json" linenums="false">

   "styles": [

>      {"input": "src/external-module/styles.scss", "lazy": true, "bundleName: "external-module"}

Reviewer 3: Can we can lazy: true to inject: false, as the former is deprecated.

See: https://github.com/angular/angular-cli/pull/14955

For example, the default asset paths can be represented in more detail using the following objects.

<code-example format="." language="json" linenums="false">

"assets": [
  { "glob": "**/*", "input": "src/assets/", "output": "/assets/" },
  { "glob": "favicon.ico", "input": "src/", "output": "/" }
]

</code-example>

You can use this extended configuration to copy assets from outside your project.
For example, the following configuration copies assets from a node package:

<code-example format="." language="json" linenums="false">

"assets": [
 { "glob": "**/*", "input": "./node_modules/some-package/images", "output": "/some-package/" },
]

</code-example>

The contents of `node_modules/some-package/images/` will be available in `dist/some-package/`.

The following example uses the `ignore` field to exclude certain files in the assets folder from being copied into the build:

<code-example format="." language="json" linenums="false">

"assets": [
 { "glob": "**/*", "input": "src/assets/", "ignore": ["**/*.svg"], "output": "/assets/" },
]

</code-example>

> {@a style-config}

Reviewer: The name would also need to be updated in all the locations that reference this anchor.

### Styles and scripts configuration

An array entry for the `styles` and `scripts` options can be a simple file references, or an object that points to an extra entry-point file.
The associated builder will load that file and its dependencies as a separate bundle during the build.

> Using this configuration you have the option of naming the bundle for the entry point and excluding it from injection. This enables [lazy loading](guide/glossary#lazy-loading), but does not implement it. You still need to implement lazy loading of the bundle if that is required.

Reviewer: Does "and/or" work here or is there some better way to indicate that you can define one or both options? Just having "and" here makes it feel like both options are required.

It would also be nice to mention that `"lazy": false` is the default.

Author: Does a bundle have to be named to be lazy-loaded? Is there any other reason to name it?

Reviewer: Lazily-injected would be the correct term if I understood the previous comments correctly. I don't believe that giving it a name is required if `"lazy": true`.

Reviewer 4: I think anything to do with lazy loading should really be a footnote.  The “lazy” option doesn’t actually cause any type of lazy loading.  It’s consistently a cause for confusion in the CLI issue tracker.

Personally, I think I would prefer if it wasn’t officially documented and the replacement (once added) be included instead.

Reviewer 3: +1, the new option should be added pretty soon anyways.

Reviewer 4: All the option really does is turn off the behavior that adds the described bundle into the application’s “index.html”.
There are also legitimate use cases for this that are not related to lazy loading. Something else to consider is that it is also not required to use the option or the complex configuration option described here to lazy load.  That aspect needs to be done manually in code no matter the method used to create a CSS or JS file.

Author: I've removed the reference to lazy loading, but retained the current name of the field. We can hold off with this until it is renamed, if that is happening soon.
Is it correct to say that the effect of `lazy=true` is to exclude the bundle from injection?
Is it correct that the example adds both the script and the style to the same bundle?

For example:

<code-example format="." language="json" linenums="false">

   "styles": [
     {"input": "src/external-module/styles.scss", "lazy": true, "bundleName: "external-module"}
     ]
   "scripts": [{
     "input": "src/external-module/main.ts", "lazy": true, "bundleName: "external-module"}
     ]

</code-example>

You can mix simple and complex file references for styles and scripts.

<code-example format="." language="json" linenums="false">

"styles": [
  "src/styles.css",
  "src/more-styles.css",
  { "input": "src/lazy-style.scss", "lazy": true },
  { "input": "src/pre-rename-style.scss", "bundleName": "renamed-style" },
],

</code-example>

> In Sass and Stylus you can also make use of the `includePaths` functionality for both component and

Reviewer: This seems to be duplicated here. It appears again below with the proper header.

project without the need for a relative path:

<code-example format="." language="ts" linenums="false">

// src/app/app.component.scss
// A relative path works
@import '../style-paths/variables';
// This also works
@import 'variables';

</code-example>

Note that you will also need to add any styles or scripts to the `test` builder if you need them for unit tests.
See also [Using runtime-global libraries inside your app](guide/using-libraries#using-runtime-global-libraries-inside-your-app).

{@ style-preprocessor}

#### Style preprocessor options

In Sass and Stylus you can also make use of the `includePaths` functionality for both component and global styles, which allows you to add extra base paths that will be checked for imports.

To add paths, use the `stylePreprocessorOptions` option:

"stylePreprocessorOptions": {
  "includePaths": [
    "src/style-paths"
  ]
},

Files in that folder, such as `src/style-paths/_variables.scss`, can be imported from anywhere in your project without the need for a relative path:

```ts
// src/app/app.component.scss
// A relative path works
@import '../style-paths/variables';
// But now this works as well

> @import 'variables';

Reviewer: It's great that this will finally be in the official docs!

```

{@ optimize-and-srcmap}

### Optimization and source map configuration

The `optimization` and `sourceMap` command options are simple Boolean flags.
You can supply an object as a configuration value for either of these to provide more detailed instruction.

* The flag `--optimization="true"` applies to both scripts and styles. You can supply a value such as the following to apply optimization to one or the other:

<code-example format="." language="json" linenums="false">

   "optimization": {"scripts": true, "styles": false}

</code-example>

* The flag `--sourceMap="true"` outputs source maps for both scripts and styles.
You can configure the option to apply to one or the other.
You can also choose to output hidden source maps, or resolve vendor package source maps.
For example:

<code-example format="." language="json" linenums="false">

>    "sourceMaps": {"scripts": true, "styles": false, "hidden":true, "vendor":true}

Reviewer: ```suggestion
   "sourceMaps": {"scripts": true, "styles": false, "hidden": true, "vendor": true}
```

</code-example>

<div class="alert is-helpful>

   When using hidden source maps, source maps will not be referenced in the bundle.

>    These are useful if you only want source map to map error stack traces in error reporting tools,

Reviewer: ```suggestion
   These are useful if you only want source maps to map error stack traces in error reporting tools,
```

>    but don't want to expose your source map in the browser tools.

Reviewer: ```suggestion
   but don't want to expose your source maps in the browser developer tools.
```

 { "glob": "**/*", "input": "src/assets/", "ignore": ["**/*.svg"], "output": "/assets/" },
]
</code-example>

### Configuring lazy-loaded styles and scripts

The `styles` and `scripts` options can accept an *extraEntryPoint* object to enable lazy loading styles or scripts.

> For example: `{"input": "src/styles/foo.css", "lazy": true, "bundleName: "foo"}`.

Reviewer: ```suggestion
Some examples: 
- `"styles": [{"input": "src/external-module/styles.scss", "lazy": true, "bundleName: "external-module"}]`
- `"scripts": [{"input": "src/external-module/main.ts", "lazy": true, "bundleName: "external-module"}]`
```

 { "glob": "**/*", "input": "src/assets/", "ignore": ["**/*.svg"], "output": "/assets/" },
]
</code-example>

> ### Style and script configuration

Reviewer 3: Plural maybe? since the options are plural.

> The `styles` and `scripts` options can accept an 'extraEntryPoint' object that points to an extra [entry point](guide/glossary#entry-point) file.

Reviewer 3: For better and more detailed descriptions I'd see https://github.com/angular/angular-cli/wiki/stories-global-styles & https://github.com/angular/angular-cli/wiki/stories-global-scripts

Author: Some of the content from those pages landed in the Libraries guide - adding a reference here.

The associated builder will load that file and its dependencies as a separate bundle during the build.

You can optionally label the bundle for such an entry point so that it can be loaded lazily.
For example:
- `"styles": [{"input": "src/external-module/styles.scss", "lazy": true, "bundleName: "external-module"}]`
- `"scripts": [{"input": "src/external-module/main.ts", "lazy": true, "bundleName: "external-module"}]`

> ### Optimization and source map configuration

Reviewer 3: Will try to write up something for this section tomorrow, as this is not entirely correct and at times it’s not very clear

Author: It's just a guess. Thanks.

> The `optimization` and `sourceMap` options to the `ng build` command are simple Boolean flags.

Reviewer 3: These options are not only for `ng build` but rather all the targets that use `sourceMaps` and `optimizaton`

You can supply an object as a configuration value for either of these options that provides more detailed instruction.

* The flag `--optimization="true"` applies to both scripts and styles. You can supply a value such as the following to apply optimization to one or the other:
   `"optimization": {"scripts": true, "styles": false}`

* The flag `--sourceMap="true"` outputs source maps for both scripts and styles.
You can configure the option to apply to one or the other.

> You can also choose to output hidden source maps for error reporting tools, or resolve vendor package source maps.

Reviewer 3: Maybe we should put an alert info box explaining what hidden sourcemaps are;

Something along
```
When using hidden source maps, source maps will not be referenced in the bundle. 

These are useful if you only want source map to map error stack traces in error reporting tools, but don't want to expose your source map in the browser tools.
```

For example:

>    `"sourceMaps": {"scripts": true, "styles": false, "hidden":true, "vendor":true}`

Reviewer 3: Can we wrap json objects inside and also format it appropriately?
```
<code-example format="." language="json" linenums="false">
</code>
```

### Configuring lazy-loaded styles and scripts

The `styles` and `scripts` options can accept an 'extraEntryPoint' object that points to an extra [entry point](guide/glossary#entry-point) file.
The associated builder will load that file and its dependencies as a separate bundle during the build.

> You can optionally label the bundle for such an entry point so that it can be loaded lazily.

Reviewer: This isn't accurate. My previous suggestion is recommended. Please ask for clarification if it is not clear.

Author: OK -- I guess it's not clear, since this is what I thought you meant. Can you restate it?

Reviewer 3: The key for lazy loading is the `lazy` option, if no bundle name is provided it will still be generated as “lazy”.
